benchmarks/spiegelman_et_al_2016_gcubed/spiegelman_material.cc

```diff
@@ -0,0 +1,680 @@
+    {
+      std::vector<double> parameter_list;
+      parameter_list = Utilities::string_to_double(Utilities::split_string_list(prm.get (parameter)));
+      if (parameter_list.size() == 1)
+        parameter_list.resize(n_fields, parameter_list[0]);
+
+      AssertThrow(parameter_list.size() == n_fields,
+                  ExcMessage("Length of " + parameter + " list (size "+ std::to_string(parameter_list.size()) +") must be either one,"
+                             " or n_compositional_fields+1 (= " + std::to_string(n_fields) + ")."));
+
+      return parameter_list;
+    }
+  }
+
+  namespace MaterialModel
+  {
+    template <int dim>
+    std::vector<double>
+    DruckerPragerCompositions<dim>::
+    compute_volume_fractions( const std::vector<double> &compositional_fields) const
+    {
+      std::vector<double> volume_fractions( compositional_fields.size()+1);
+
+      // clip the compositional fields so they are between zero and one
+      std::vector<double> x_comp = compositional_fields;
+      for ( unsigned int i=0; i < x_comp.size(); ++i)
+        x_comp[i] = std::min(std::max(x_comp[i], 0.0), 1.0);
+
+      // sum the compositional fields for normalization purposes
+      double sum_composition = 0.0;
+      for ( unsigned int i=0; i < x_comp.size(); ++i)
+        sum_composition += x_comp[i];
+
+      if (sum_composition >= 1.0)
+        {
+          volume_fractions[0] = 0.0;  // background material
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1]/sum_composition;
+        }
+      else
+        {
+          volume_fractions[0] = 1.0 - sum_composition; // background material
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1];
+        }
+      return volume_fractions;
+    }
+
+    template <int dim>
+    double
+    DruckerPragerCompositions<dim>::
+    compute_second_invariant(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const
+    {
+      const double edot_ii_strict = std::sqrt(strain_rate*strain_rate);
+      const double edot_ii = std::max(edot_ii_strict, min_strain_rate*min_strain_rate);
+      return edot_ii;
+    }
+
+    template <int dim>
+    double
+    DruckerPragerCompositions<dim>::
+    compute_viscosity(const double edot_ii,const double pressure,const int comp,const double prefactor,const bool regularize, const double min_visc, const double max_visc) const
+    {
+      double viscosity;
+      if (comp == 0)
+        {
+          const double strength = ( (dim==3)
+                                    ?
+                                    ( 6.0 * cohesion[comp] * cos_phi[comp] + 6.0 * std::max(pressure,0.0) * sin_phi[comp] )
+                                    / ( sqrt_3 * ( 3.0 + sin_phi[comp] ) )
+                                    :
+                                    cohesion[comp] * cos_phi[comp] + std::max(pressure,0.0) * sin_phi[comp] );
```

Comment:
You are here in a block where you know that `comp==0`, so we only ever use the first element of `cohesion`, `cos[phi]`, etc. But below, when `comp != 0`, we don't actually references these quantities at all and only return the prefactor. Is this a mistake?

Reply:
Good point. It has grown a bit like this since it was first intended as a Drucker Prager material model which supports different compositions, but then for the Spiegelman benchmark I needed a constant viscosity. I now changed it so that when the prefactor (which I renamed to constant_viscosity) is zero, it uses the Drucker Prager rhelology, otherwise it uses that value as the viscosity.